crypto/cmp/cmp_util.c

```diff
@@ -155,12 +155,27 @@ void OSSL_CMP_print_errors_cb(OSSL_CMP_log_cb_t log_fn)
     while ((err = ERR_get_error_all(&file, &line, &func, &data, &flags)) != 0) {
         const char *component =
             improve_location_name(func, ERR_lib_error_string(err));
```

Comment:
...

Reply:
As commented there:
```
/* this is similar to ERR_print_errors_cb, but uses the CMP-specific cb type */
```

Reply:
I remember that we've had this discussion before...  I've also recently seen the output, and was quite confused to begin with, 'cause I couldn't make sense of what I was looking at.

I find it *problematic* that the different apps output errors in varying ways.  Mind you, though, that I'm not terribly fond of the default output, but at the very least, I've become accustomed to it and know how to read it...  so when I'm suddenly faced with a very different format, I get thrown off.

If you want to reform how errors are output, that should be taken to a more *global* OpenSSL level, not piece meal per app.

BTW, remind me, why couldn't ERR_print_error_cb be used?  All you would have to do is provide the callback instead of replace it completely...

Reply:
For the sake of this PR, though, this discussion should probably be brought elsewhere.  I'll try to remember to raise an issue.

Reply:
> I remember that we've had this discussion before... 

Yes. Part of the discussion was, as far as I recall, that I believe that it would be good to 
* introduce a severity level
* unify the error reporting via the error queue with logging

Both of this is done in `cmp_util.c`, but (so far) not for the rest of the library.
  
>I've also recently seen the output, and was quite confused to begin with, 'cause I couldn't make sense of what I was looking at.

Maybe you saw a mix of the usual output, which contains certain (mostly) needless detail and partly unordered pieces of information like this:
```
C0F706403F7F0000:error:80000002:system library:file_open:No such file or directory:providers/implementations/storemgmt/file_store.c:278:calling stat(t.crt)
```
with differently formatted CMP-specific error output (which by itself IMHO is more logical and readable), such as
```
CMP:providers/implementations/storemgmt/file_store.c:278:file_open:error:No such file or directory:calling stat(t.crt)
```
 
> I find it _problematic_ that the different apps output errors in varying ways.

I agree - that's why I had general adaptations as outlined above, but at that point it was decided against them.
 
> Mind you, though, that I'm not terribly fond of the default output, but at the very least, I've become accustomed to it and know how to read it... so when I'm suddenly faced with a very different format, I get thrown off.

I understand.

> If you want to reform how errors are output, that should be taken to a more _global_ OpenSSL level, not piece meal per app.

Yes, we should go for that.

Yesterday evening I had already started exploring ways to unify the implementation in `cmp_util.c` with the one in `err{_prn}.c`, and I am going to provide a PR that at least removes much of the current ugly code duplication.
 
> BTW, remind me, why couldn't ERR_print_error_cb be used? All you would have to do is provide the callback instead of replace it completely...

It's not that easy because the parameters of the callback are quite different:
```
int (*cb) (const char *str, size_t len, void *u)
```
vs.
```
int (*OSSL_CMP_log_cb_t)(const char *func, const char *file, int line, OSSL_CMP_severity level, const char *msg)
```
Yet since meanwhile the callback used on `ERR_print_errors_cb()` has the general-purpose  `void *u` parameter it should be possible to map the `OSSL_CMP_log_cb_t` to it.

> For the sake of this PR, though, this discussion should probably be brought elsewhere. I'll try to remember to raise an issue. I suggest discussing all this further in the issue that you are going to file.

Yes, this would be better.